### PR TITLE
DM-38461: Handle optional fields in type annotations.

### PR DESCRIPTION
**Pull request by TallJimbo**
## Checklist

- [ ] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: python/lsst/pex/config/config.py [hunk 1]
```diff
@@ -453,20 +466,30 @@ def _parseTypingArgs(
...
+                case [types.NoneType, annotated_dtype]:
+                    pass
+                case [annotated_dtype, types.NoneType]:
+                    pass
+                case _:
+                    raise ValueError("Unsupported union type for Field.")
+            annotated_optional = True
+        else:
+            annotated_dtype = unpackedParams
+            annotated_optional = False
```
Comment (natelust) [inline python/lsst/pex/config/config.py:487]: Consider this:
```suggestion
        match (get_origin(unpackedParams) is types.UnionType, get_args(unpackedParams)):  #  noqa: E721
            case (True as annotated_optional, (types.NoneType, annotated_dtype)):
                pass
            case (True as annotated_optional, (annotated_dtype, types.NoneType)):
                pass
            case (types.UnionType, _):
                raise ValueError("Unsupported union type for Field.")
            case (annotated_optional, _):
                annotated_dtype = unpackedParams
```

File: tests/test_Config.py [hunk 2]
```diff
@@ -136,6 +136,40 @@ def testFieldTypeAnnotationRuntime(self):
         testField = pexConfig.Field[int](doc="", dtype=int)
         self.assertEqual(testField.dtype, int)
 
+        # check interplay between optional=True and Unions.
+        testField = pexConfig.Field[int | None](doc="")
```
Comment (natelust) [inline tests/test_Config.py:140]: can you add a test with `[None | int]`

File: python/lsst/pex/config/config.py (hunk 1 above)
Comment (natelust) [reply python/lsst/pex/config/config.py:487]: Also you could import UnionType to shorten line 476

**Review by natelust — COMMENTED** [review]
You also need to get the other overrides of _parseTypingArgs, like in dictField and configurableField, at the time it was easier to give it its own implementation

**Comment by natelust** [comment]
I also don't remember what the thought process was on configurableField, its implementation might just attempt to punt on it all

**Comment by natelust** [comment]
Since you are removing the 3.8 code paths can you get this as well: https://github.com/lsst/pex_config/pull/96/files#diff-994e6b4ebd89599523c436e976eb214dda8cf47d8be0573211481567722908f3R67

**Review by TallJimbo — COMMENTED** [review]
So, `ConfigurableField` does indeed punt, and it doesn't support `optional=True` anyway, so there's nothing to do there.

It looks like `DictField` ideally should get better `optional=True` type annotations, but I think it's impossible, because it inherits from `Field[Dict[KeyTypeVar, ItemTypeVar]]`, and we can't change that to `Field[Dict[KeyTypeVar, ItemTypeVar] | None]` later at construction time.  Changing the generic parameters for `DictField` would be the only way to fix that, but it'd be a worse typing interface when `optional=False` (as it almost always is for this field) as well as a breaking change.  So I'm inclined to just add a note to the docs for the `optional` constructor argument saying that if you pass `True` you may have to tell mypy to ignore some lines.

File: python/lsst/pex/config/config.py (hunk 1 above)
Comment (natelust) [reply python/lsst/pex/config/config.py]: I'm fine if you want to leave it, it definitely is in the area of different people  will find different things readable. But I reserve the right to highlight this in the future when you become a match operator aficionado.

For completeness, this is saying something like match ( (true or false), args), in the case it is a union type, either of the first two cases will match. The third case statement I forgot to update from a previous version I was going to suggest, and it too would be true.

in the case of the last one, you are simply saying bind the match argument to a variable no matter what it is, and we know that True would have matched above so the only thing it could be is false, so annotated_optional gets set as False. It is basically just the catch all _ that takes advantage of an assigment at the same time.

**Comment by natelust** [comment]
Let me look at DictField again to see if I see any other option, but I am happy with everything else.
